### DISCUSSION TITLE
Please make the right-click menu item "Share" available on macOS

### DISCUSSION THREAD

**Discussion by ghost** [doublecmd/doublecmd#1265] [Suggestions]
On macOS, please make the right-click menu item "Share" available (like in Finder).

**Comment by rich2014** [comment]
good idea.

however, there are a lot of bug fixes for Lazarus/Cocoa/MacOS recently, many of which are also helpful to DC, so these bug fixes have a higher priority.

so it will take a while before i have time to implement the `Share` function.

**Comment by ghost** [comment]
Sure. Thank you for considering my wish, and for your good work in general.

**Comment by ags1234** [comment]
> there are a lot of bug fixes for Lazarus/Cocoa/MacOS recently

All made by mr. rich2014 :)

**Comment by Skif-off** [comment]
Are there commands for this in macOS?
I thought about [file associations](https://doublecmd.github.io/doc/en/configuration.html#ConfigAssociations) and the "Actions" submenu, but I'm not very familiar with this OS.

**Comment by ghost** [comment]
File associations is not the right way, I would say.

"Share" (in German: "Teilen") is an own entry in the right-click menu. See screenshot below.

![screenshot](https://github.com/doublecmd/doublecmd/assets/79788982/5d65c557-0ef7-47f6-9aef-c2d50901fd8a)

**Comment by Skif-off** [comment]
@Andreas0602,
the "File Associations" settings section allows not only to configure or change associations with applications (F3, F4, Enter/double click): you can also add any actions by file type or for all (see special values `file` and `folder`) or set icons without changing associations.
[Example](https://doublecmd.github.io/doc/en/images/imgDC/pic49.png) of an additional action.

But I don't know if there is a suitable command or application in macOS for which user can specify a path as a parameter.

**Comment by ghost** [comment]
Right, I see. So the Actions submenu would be okay (though I still think an entry in the first-level menu would be better — like in Finder, ForkLift, or MaxCommander).

I don't know either whether it's possible to perform the Share functionality by simply invoking an existing macOS command. But I'm afraid it's not that easy, because there aren't many other file managers on the Mac that offer this feature.

**Comment by rich2014** [comment]
@Andreas0602 

`Share` ContextMenu supported on MacOS.

download the snapshot please.

**Comment by ghost** [comment]
Thank you very much, @rich2014! Just two comments:

1. The name of the menu entry should be "Share ...", not "Share".
2. The behavior is a bit different than in Finder, ForkLift, or MaxCommander. When you click "Share ..." in these apps and then select the Messages app, the Messages app will start and you can then add some text to your message and finally send it. In contrast to that, with DC, you cannot add any text to your message. That's a disadvantage. Would it be possible to improve that?

**Comment by Skif-off** [comment]
@Andreas0602

1. "..." usually indicates that the menu item opens some kind of window and some kind of user interaction is required. Or the items in the submenu are a continuation of the submenu name, i.e.
```
"Share..." > "in email"
             "in social network"
              and so on
```
[Screenshots](https://www.google.com/search?q=macos+context+menu+share&tbm=isch).
Or I didn't understand something.

**Comment by ghost** [comment]
In Apple's Finder, the name is "Share...".

![Finder](https://github.com/doublecmd/doublecmd/assets/79788982/b9a445b3-87f9-4339-90a6-b032708760a2)

**Reply by alexx2000** [reply]
It depends on macOS [version](https://lapcatsoftware.com/articles/VenturaShare.html).

**Comment by rich2014** [comment]
1. referring to MacOS 13, `share` became `share...`

2. @Andreas0602 

`text` can be added to Messages on MacOS 12.6.

could you provide more detailed information?

<img width="442" alt="image" src="https://github.com/doublecmd/doublecmd/assets/102933068/a105b29d-fbeb-45a0-9bdf-32aaa6b59fce">

**Comment by Skif-off** [comment]
@Andreas0602

If I understood everything correctly, "Share" _submenu_ has been added to the DC’s context menu.
But in your screenshot, "Share..." is not a _submenu_, there are only *two* _submenus_: "Open With" and "Quick Actions". I don't really understand why you don't see the difference in implementation :)

This is probably not a problem, but it looks strange.

**Comment by ghost** [comment]
@rich2014, on my system, it looks the same as in your screenshot. I was just not aware that you can type text into the grey area. I expected a window with a white background. So, from a functional point of view, your implementation is okay -- with one small drawback: With your implementation, the complete text will always appear **before** the image. In contrast to that, with Apple's implementation, you can place text **before** and/or **after** the image as you wish.

@Skif-off, I didn't want to make a big deal out of the "Share" versus "Share..." topic. I just wanted to point it out, in case it was missed. If you insist on "Share", I have no problem with that.

**Comment by rich2014** [comment]
@Andreas0602

only native api of MacOS are called.

could you provide the screenshot?

**Comment by ghost** [comment]
@rich2014

When I right-click an image file in **Double Commander**, then click the "Share..." menu item, and then click "Messages",  the window for composing the message will look like this:

![Sending with Double Commander](https://github.com/doublecmd/doublecmd/assets/79788982/7b834d39-e836-4b19-bd0d-fd83af537518)

The sent message in the Messages app looks like this:

![Message sent with Double Commander, shown in Messages app](https://github.com/doublecmd/doublecmd/assets/79788982/ddac3388-7b6a-4ac5-a93a-c746aee82f7a)

-----------------------------

When I right-click an image file in **ForkLift**, then click the "Share..." menu item, and then click "Messages", the window for composing the message (in the Messages app) will look like this:

![Sending with ForkLift](https://github.com/doublecmd/doublecmd/assets/79788982/83a48fb9-e1c0-4207-8784-ebe80e694b07)

The sent message in the Messages app looks like this:

![Message sent with ForkLift, shown in Messages app](https://github.com/doublecmd/doublecmd/assets/79788982/8112edeb-7d21-419b-8325-18606113f198)

**Comment by alexx2000** [comment]
> When I right-click an image file in ForkLift, then click the "Share..." menu item

What about **Finder**? ForkLift is a third party application.

**Comment by ghost** [comment]
The behavior with Finder is exactly the same as the behavior with ForkLift.

**Comment by ghost** [comment]
I only took ForkLift as an example to point out that it is possible for a third-party app to implement exactly the same behavior as in Finder.

**Comment by rich2014** [comment]
tested on MacOS12 and 13 respectively, there are some subtle differences

1. on MacOS 12, DC behaves the same as Finder

2. on MacOS 13, DC behaves the same as on MacOS12, but Finder behaves differently than on MacOS12

3. on MacOS 13, Finder behaves the same as the ScreenShot provided by @Andreas0602

it is estimated that this is not caused by the code itself, but may be related to signatures and permissions.

**Comment by ghost** [comment]
I just noticed another shortcoming of the "Share > Messages" dialog used by DC: The keyboard shortcuts command+C and command+V do not work. (Tested with doublecmd-1.2.0-11560.cocoa.aarch64 on macOS 15.0.)

@rich2014, It would be great if you could somehow manage to make DC use the same "Share > Messages" dialog as Finder, ForkLift, and DoubleXplorer (where command+C and command+V work properly).

**Comment by rich2014** [comment]
@Andreas0602 

try the snapshoot please.

this is not a problem with sharing itself, but rather that these dialog boxes rely on the ShortCut provided by the menu.

by default, DC does not create a menu with corresponding editing shortcuts. now the problem is solved by creating an additional editing menu on macOS.

**Reply by ghost** [reply]
command+C and command+V work fine now. Thank you very much for that, @rich2014!

**Reply by rich2014** [reply]
it's great!
